File: 코드/ch4/clip3/boj15681/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Main {
    static List<Integer>[] tree;
    static boolean[] check;
    static int[] sum;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), r = sc.nextInt(), q = sc.nextInt();
        check = new boolean[n + 1];
        sum = new int[n + 1];
        tree = new ArrayList[n + 1];
        for (int i = 0; i < tree.length; i++) {
            tree[i] = new ArrayList<>();
        }
        for(int i=0; i<n-1; i++) {
            int u = sc.nextInt(), v = sc.nextInt();
            tree[u].add(v);
            tree[v].add(u);
        }
        sum[r] = getSum(r);
        for (int i = 0; i < q; i++) {
            int u = sc.nextInt();
            System.out.println(sum[u]);
        }
    }

    public static int getSum(int node) {
        if (sum[node] != 0) return sum[node];
        check[node] = true;
        int result = 1; // 자기 자신
        for (int child : tree[node]) {
            if (!check[child]) {
                result += getSum(child);
            }
        }
        return sum[node] = result;
    }
}

File: 코드/ch4/clip3/boj15681/main.cpp
#include <iostream>
#include <vector>

std::vector<int> tree[100001];
int sum[100001];
bool check[100001];

int getSum(int nodeNumber)
{
    if(sum[nodeNumber] != 0)
    {
        return sum[nodeNumber];
    }

    check[nodeNumber] = true;
    sum[nodeNumber] = 1;

    for(int child : tree[nodeNumber])
    {
        if(false == check[child])
        {
            sum[nodeNumber] += getSum(child);
        }
    }

    return sum[nodeNumber];
}

int main()
{
    int N, R, Q, U, V;
    scanf("%d %d %d", &N, &R, &Q);
    for(int i = 0; i < N - 1; i++)
    {
        scanf("%d %d", &U, &V);
        tree[U].push_back(V);
        tree[V].push_back(U);
    }

    getSum(R);

    for(int i = 0; i < Q; i++)
    {
        scanf("%d", &U);
        printf("%d\n", sum[U]);
    }

    return 0;
}